General improvements

theseal opened the pull request:
* Possibility to reload TLS certificate without restart (listening for HUP)
* Log all proxies received in the X-Forwarded-For header(s)
* Graceful shutdown - allowing all clients to finish their request before shutting down (don't worry, we have a timeout)
* The log now includes more useful information when debugging. Like hostname and which build is running

Maybe each point should have been it's own PR 🤷‍♂️

theseal committed:
Improve logging

Added useful information for debugging

Closes #5

theseal committed:
Log all proxies

If available, show all proxies from the request, not only the latest.

Closes #4

theseal committed:
Allow clients to forfil their request before exit

aka graceful shutdown

Closes #6.

theseal committed:
Possibilty to reload TLS key pair without restart

theseal committed:
Consistency

theseal committed:
Can't be reached any more

Replaced by the certStore

theseal committed:
Not in use any more

eest commented on build/publisher.go:
```diff
@@ -30,13 +37,22 @@ func (m *myMux) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+			adresses := strings.Split(header, ",")
+			for _, adress := range adresses {
+				adress := strings.TrimSpace(adress)
+				merged_xffs = append(merged_xffs, adress)
```
We never verify that the string is a valid IP address. Not sure this is needed but worth considering.

theseal replied:
Not saying it's a good answer but we didn't do it before either. With that said I think it would be good to have some sort of protection since we actually parse the logs looking for addresses.

It is possible that HAproxy and/or apache adds some protection here but I know that others run this server with out a proxy aswell.

Created #11  to fix this later since I really want to the other things merged and tested.

theseal replied:
The fix was to easy to not include in this PR 🤦‍♂️

eest commented on build/publisher.go:
```diff
@@ -182,6 +249,32 @@ func main() {
+		httpServerCertStore, err := newCertStore(serverCert, serverKey)
+		if err != nil {
+			zlog.Fatal().Err(err).Msg("Unable to load x509 HTTP server cert from disk")
+			os.Exit(1)
```
Dont think we need os.Exit(1) here as zlog.Fatal() should stop the program via os.Exit(1) for us.

theseal replied:
Fixed, thanks.

theseal committed:
zlog.Fatal() handles the exit

Thanks for pointing that out @eest

theseal committed:
No formating is happening

Pointed out by @eest

theseal commented on build/publisher.go:
```diff
@@ -182,6 +249,32 @@ func main() {
+			signal.Notify(sigHup, syscall.SIGHUP)
+
+			for range sigHup {
+				zlog.Info().Msgf("HUP received")
```
Done and done 🙏

theseal committed:
Duplicated error(s)

theseal committed:
Cleaner code

theseal commented on build/publisher.go:
```diff
@@ -182,6 +249,32 @@ func main() {
+				zlog.Info().Msgf("HUP received")
+				err = httpServerCertStore.loadCert()
+				if err != nil {
+					zlog.Error().Err(err).Msg("Unable to reload x509 HTTP server cert from disk")
```
Thanks.

theseal committed:
Make sure to only log valid X-Forwarded-For

Valid as containing a real IP address

Closes #11

theseal committed:
netip is more modern

> Compared to the net.IP type, Addr type takes less memory, is immutable, and is comparable (supports == and being a map key).

eest reviewed (approved):
LGTM

btmattsson reviewed (approved):
LGTM
